batchproducer/batchproducer.go

```diff
@@ -223,15 +235,44 @@ func (b *batchProducer) run() {
 	}
 }
 
-// from/for interface BatchProducer
+// from/for interface Producer
 func (b *batchProducer) Stop() error {
-	// TODO: Immediately stop accepting new records by Add, then block until all the records in the buffer have been sent
 	if b.isRunning() {
 		b.stop <- true
 	}
 	return nil
 }
 
+// from/for interface Producer
+// TODO: send all batches in parallel, will require broader refactoring
+func (b *batchProducer) Flush(timeout time.Duration) (int, error) {
+	b.Stop()
+
+	timer := time.NewTimer(timeout)
+	if timeout == 0 {
+		timer.Stop()
+	}
+
+	timedOut := false
+
+loop:
+	for len(b.records) > 0 {
+		select {
+		case <-timer.C:
+			timedOut = true
+			break loop
```

Comment:
The [spec](https://golang.org/ref/spec#Break_statements) says:

> If there is a label, it must be that of an enclosing "for", "switch", or "select" statement, and that is the one whose execution terminates.

So in fact if the timeout occurs, we break out of the `for` loop. The idea is to then return as fast as possible. So in that case, we don’t send a final batch of stats; we just return.

If, on the other hand, we exited the loop *without* timing out, that means that we successfully sent all the buffered records and still perhaps have a little time to spare. So we go ahead and send the final batch of stats.

Does that help?